Filewatcherd-TypeScript/src/lib/IndividualFileWatchService.ts

```diff
@@ -0,0 +1,268 @@
+                log.error("Project '" + projectId + "' was asked to watch a directory, which is not supported: "
+                    + newPath);
+                continue;
+            }
+
+            paths.push(newPath);
+        }
+
+        if (paths.length === 0) {
+            return;
+        }
+
+        let mapUpdated = false;
+
+        const currProjectState = this._filesToWatchMap.get(projectId);
+
+        if (!currProjectState) {
+
+            // This is a new project we haven't seen.
+
+            const newFiles = new Map<string, PollEntry>();
+
+            for (const path of paths) {
+                const pollEntry = new PollEntry(PollEntryStatus.RECENTLY_ADDED, path, 0);
+                newFiles.set(pollEntry.absolutePath, pollEntry);
+                log.info("Files to watch - recently added for new project: " + path);
+            }
+            mapUpdated = true;
+
+            this._filesToWatchMap.set(projectId, newFiles);
+        } else {
+            // This is an existing project with at least one file we are currently monitoring.
+
+            for (const path of paths) {
+
+                const pe = currProjectState.get(path);
+                if (!pe) {
+
+                    log.info("Files to watch - recently added for existing project: " + path);
+                    currProjectState.set(path, new PollEntry(PollEntryStatus.RECENTLY_ADDED, path, 0));
+                    mapUpdated = true;
+
+                } else {
+                    // Ignore: the path is in both maps -- no change.
+                }
+            }
+
+            const pathsInParam = new Set<string>();
+            for (const path of paths) {
+                pathsInParam.add(path);
+            }
+
+            const keysToRemove = [];
+
+            // Look for values that are in curr project state, but not in the parameter list. These are
+            // files that we WERE watching, but are no longer.
+            for (const [pathInCurrentState, _] of currProjectState) {
+
+                if (!pathsInParam.has(pathInCurrentState)) {
+                    keysToRemove.push(pathInCurrentState);
+                    log.info("Files to watch - removing from watch list: " + pathInCurrentState);
+                    mapUpdated = true;
+                }
+            }
+            for (const keyToRemove of keysToRemove) {
+                currProjectState.delete(keyToRemove);
+            }
+
+            // If we're not watching anything anymore, remove the project from the state list.
+            if (currProjectState.size === 0) {
+                this._filesToWatchMap.delete(projectId);
+            }
+
+        } // end existing project else
+
+    } // end method
+
+    public dispose() {
+        if (this._disposed) {
+            return;
+        }
+        this._disposed = true;
+
+        this._filesToWatchMap.clear();
+
+    }
+
+    private threadRun() {
+
+        if (this._timer) {
+            clearInterval(this._timer);
```

Comment:
It may not matter in practice, but since you created the timer using `setTimeout` not `setInterval`, you should clear it with `clearTimeout`.

Reply:
Good catch... I have some vague memory of looking up this exact thing and thiking what I had was correct, but you are clearly right.

Fixed and reran the automated tests.